### PR TITLE
779 handle approve reject expressions

### LINKED ISSUE #779: 2 - [`x/act` V2]: Add handling for the new expression fields (Approve/Reject) in `MsgNewAction`
1. Extend `RulesRegistry` with Approve/Reject expressions.
2. Set expressions within `MsgNewAction` handling.
3. Update registry initialization in all the dependent modules, and add missing template ids.

### PR DESCRIPTION
**Pull request by mn13**
## Summary by CodeRabbit

- **New Features**
	- Simplified approval and rejection expressions for actions.
	- Enhanced event messages with specific rule IDs for approvals and rejections.
	- Introduced new fields for managing approval and rejection rules in various messages.
	- Updated commands to allow users to specify expected approval and rejection expressions when creating actions.
	- Added new methods to clearly distinguish between approval and rejection actions for owner and key management.
	- Improved rules management with methods that return both approval and rejection rules.

- **Bug Fixes**
	- Improved validation checks for rule IDs during updates to keys and spaces.
	- Enhanced error handling with new error messages for unmatched approval and rejection expressions.

### CODE REVIEW

File: proto/warden/warden/v1beta3/space.proto [hunk 1]
```diff
@@ -40,5 +40,53 @@ message Space {
...
+  // If not specified, the default Rule is used.
+  //
+  // Sign operations are:
+  // - warden.warden.Msg.NewKeyRequest
+  // - warden.warden.Msg.NewSignRequest
+  // - warden.warden.Msg.UpdateKey
+  //
+  // The default Rule is to allow any operation when at least one of its
+  // owner approves it.
+  uint64 reject_sign_rule_id = 11;
```
Comment (backsapc) [inline proto/warden/warden/v1beta3/space.proto:90]: I have an idea to merge approve & reject rules into single struct (like `RulePair`) @Pitasi

File: warden/x/act/keeper/actions.go [hunk 2]
```diff
@@ -134,45 +136,74 @@ type actionCreatorKey struct{}
...
+	if err != nil {
+		return nil, err
+	}
+
+	preprocessedRejectExpr, rejectMentions, err := k.preprocessRule(ctxWithMsg, rejectRule)
 	if err != nil {
 		return nil, err
 	}
 
+	mentions := approveMentions
```
Comment (backsapc) [inline warden/x/act/keeper/actions.go]: I'd suggest extracting this into a function/method. Lines: 174 - 187

File: warden/x/act/keeper/actions.go [hunk 3]
```diff
@@ -134,45 +136,74 @@ type actionCreatorKey struct{}
...
-		Mentions:      mentions,
-		Msg:           wrappedMsg,
-		Creator:       creator,
-		TimeoutHeight: timeoutHeight,
-		CreatedAt:     timestamp,
-		UpdatedAt:     timestamp,
+		Status:            types.ActionStatus_ACTION_STATUS_PENDING,
+		Approvers:         nil,
+		Rule:              approveRule,
+		Mentions:          mentions,
```
Comment (backsapc) [inline warden/x/act/keeper/actions.go:253]: Do we really want to keep it as is? Why don't add another `RejectMentions`? Just curious

File: warden/x/warden/keeper/rules.go [hunk 4]
```diff
@@ -19,114 +19,240 @@ func (k Keeper) RegisterRules(reg *acttypes.RulesRegistry) {
...
+
+func (k Keeper) getRejectNewKeyRequestRule(ctx context.Context, space v1beta3.Space) (acttypes.Rule, error) {
+	if space.RejectSignRuleId > 0 {
+		return k.actKeeper.GetRule(ctx, space.RejectSignRuleId)
+	} else {
+		return space.RuleRejectNewKeyRequest(), nil
+	}
+}
+
+func (k Keeper) getApproveUpdateKeyRule(ctx context.Context, space v1beta3.Space, key v1beta3.Key) (acttypes.Rule, error) {
```
Comment (backsapc) [inline warden/x/warden/keeper/rules.go:224]: Unused `key v1beta3.Key`

File: warden/x/warden/keeper/rules.go [hunk 5]
```diff
@@ -19,114 +19,240 @@ func (k Keeper) RegisterRules(reg *acttypes.RulesRegistry) {
...
+	if space.ApproveSignRuleId > 0 {
+		return k.actKeeper.GetRule(ctx, space.ApproveSignRuleId)
+	} else {
+		return space.RuleApproveUpdateKey(), nil
+	}
+}
+
+func (k Keeper) getRejectUpdateKeyRule(ctx context.Context, space v1beta3.Space, key v1beta3.Key) (acttypes.Rule, error) {
```
Comment (backsapc) [inline warden/x/warden/keeper/rules.go:234]: Unused `key v1beta3.Key`

File: warden/x/warden/keeper/rules.go [hunk 6]
```diff
@@ -19,114 +19,240 @@ func (k Keeper) RegisterRules(reg *acttypes.RulesRegistry) {
...
 	}
 
 	space, err := k.SpacesKeeper.Get(ctx, key.SpaceId)
 	if err != nil {
-		return acttypes.Rule{}, err
+		return acttypes.Rule{}, acttypes.Rule{}, err
 	}
 
-	rule, err := k.getKeyRule(ctx, space, key)
+	approveRule, err := k.getApproveUpdateKeyRule(ctx, space, key)
```
Comment (backsapc) [inline warden/x/warden/keeper/rules.go:125]: It was `rule, err := k.getKeyRule(ctx, space, key)`, I think you don't need `getApproveUpdateKeyRule`. You need to replace this by  `getApproveKeyRule`

File: warden/x/warden/keeper/rules.go (hunk 4 above)
Comment (backsapc) [reply warden/x/warden/keeper/rules.go:224]: Looks like we don't need this method

**Review by backsapc — CHANGES_REQUESTED** [review]
Overall good, there's a couple of comments for `rules.go`. Also I think it may be a good idea to add  `RulePair` structure, because they're linked, imho. But the current design also allows us to have single reject rule for all other commands..

File: warden/x/act/keeper/actions.go [hunk 7]
```diff
@@ -35,6 +36,7 @@ var _ shield.Environment = ApproversEnv{}
 // result in the database.
 func (k Keeper) TryExecuteAction(ctx context.Context, act *types.Action) error {
 	ready, err := k.checkActionReady(ctx, *act)
+	log.Printf("\nready %v\n", ready)
```
Comment (artur-abliazimov) [inline warden/x/act/keeper/actions.go]: Is it for testing only? Should we remove it?

File: warden/x/act/keeper/actions.go [hunk 8]
```diff
@@ -134,45 +136,74 @@ type actionCreatorKey struct{}
...
 
 	ctx = ctxWithActionCreator(sdk.UnwrapSDKContext(ctx), creator)
-	ctx, rule, err := k.rulesRegistry.Get(ctx, msg)
+	ctx, approveRule, rejectRule, err := k.rulesRegistry.Get(ctx, msg)
 	if err != nil {
 		return nil, errors.Wrapf(types.ErrNoRuleRegistryHandler, "%v", err)
 	}
 
-	// todo: check that expressions from rulesRegistry (templateRegistry) match with expected
+	if approveRule.Expression.String() != expectedApproveExpression.String() {
```
Comment (artur-abliazimov) [inline warden/x/act/keeper/actions.go]: Would the AST Node comparison with `DeepEqual` be more appropriate here in terms of performance (no ordinal string comparisons for a large AST) and correctness?

Example:
```
if !reflect.DeepEqual(approveRule.Expression, expectedApproveExpression) {
    return nil, types.ErrApproveExpressionNotMatched
}
```

File: warden/x/warden/keeper/rules.go (hunk 6 above)
Comment (mn13) [reply warden/x/warden/keeper/rules.go:125]: I think we accidentally didn't use updateKeyRule

File: warden/x/act/keeper/actions.go [hunk 9]
```diff
@@ -134,45 +136,74 @@ type actionCreatorKey struct{}
 // AddAction creates a new action.
 // The action is created with the provided creator as the first approver.
 // This function also tries to execute the action immediately if it's ready.
-func (k Keeper) AddAction(ctx context.Context, creator string, msg sdk.Msg, timeoutHeight uint64, expectedApproveExpression ast.Expression, expectedRejectExpression ast.Expression) (*types.Action, error) {
+func (k Keeper) AddAction(ctx context.Context, creator string, msg sdk.Msg, timeoutHeight uint64, expectedApproveExpression *ast.Expression, expectedRejectExpression *ast.Expression) (*types.Action, error) {
```
Comment (artur-abliazimov) [inline warden/x/act/keeper/actions.go:206]: Lets merge the latest changes from the main branch. The call below: 
```
if err := k.TryExecuteAction(ctx, act); err != nil {
	return nil, err
}
```

must be replaced with:
```
if err := k.TryExecuteVotedAction(ctx, act); err != nil {
	return nil, err
}
```

File: warden/x/warden/keeper/rules.go (hunk 6 above)
Comment (mn13) [reply warden/x/warden/keeper/rules.go:125]: The difference with getApproveKeyRule/getRejectKeyRule in returning RuleApproveUpdateKey/RuleRejectUpdateKey in last else branch. RuleUpdateKey not used before changes, I think it's a mistake.

File: warden/x/act/keeper/actions.go (hunk 3 above)
Comment (mn13) [reply warden/x/act/keeper/actions.go:253]: Well, user can check his mention in both aprrove/reject rules itself, the fact that address is mentioned matters

File: warden/x/act/keeper/actions.go (hunk 8 above)
Comment (mn13) [reply warden/x/act/keeper/actions.go]: Makes sense. Didn't know about DeepEqual in go, thanks!

File: warden/x/warden/keeper/rules.go [hunk 10]
```diff
@@ -19,114 +19,240 @@ func (k Keeper) RegisterRules(reg *acttypes.RulesRegistry) {
...
 	}
 
 	space, err := k.SpacesKeeper.Get(ctx, key.SpaceId)
 	if err != nil {
-		return nil, acttypes.Rule{}, err
+		return nil, acttypes.Rule{}, acttypes.Rule{}, err
 	}
 
-	rule, err := k.getKeyRule(ctx, space, key)
+	approveRule, err := k.getApproveKeyRule(ctx, space, key)
```
Comment (artur-abliazimov) [inline warden/x/warden/keeper/rules.go]: Hm, I think that this `getApproveKeyRule` is misleading in the `newSignRequestRule` method. Can we rename it to be more related to signature? Like, `getApproveSignatureRule` or something similar.

File: warden/x/act/keeper/actions.go [hunk 11]
```diff
@@ -202,45 +203,61 @@ type actionCreatorKey struct{}
...
```
Comment (artur-abliazimov) [inline warden/x/act/keeper/actions.go:206]: Also, this code must be replaced:

```
// TODO AT: Remove after full switch to Votes
if err := act.AddOrUpdateVote(sdkCtx, creator, types.ActionVoteType_VOTE_TYPE_APPROVED); err != nil {
	return nil, err
}
```

with this one:
```
if err := act.AddOrUpdateVote(sdkCtx, creator, types.ActionVoteType_VOTE_TYPE_APPROVED); err != nil {
	return nil, err
}
```

Comment (artur-abliazimov) [reply warden/x/act/keeper/actions.go:206]: I see that you removed the `AddApprover`, but it must be replaced with `AddOrUpdateVote` instead.

**Comment by mn13** [comment]
I'm fixing tests

File: warden/x/act/keeper/actions.go (hunk 11 above)
Comment (mn13) [reply warden/x/act/keeper/actions.go:206]: It's already there after rebase, at :265

Comment (artur-abliazimov) [reply warden/x/act/keeper/actions.go:206]: Or, I see it now. Thanks!

File: warden/x/warden/keeper/rules.go (hunk 6 above)
Comment (backsapc) [reply warden/x/warden/keeper/rules.go:125]: @Pitasi we need your opinion

File: warden/x/warden/keeper/rules.go (hunk 10 above)
Comment (backsapc) [reply warden/x/warden/keeper/rules.go]: @artur-abliazimov  Isn't used in `Keys` related methods too?

File: warden/x/warden/keeper/rules.go [hunk 12]
```diff
@@ -19,114 +19,244 @@ func (k Keeper) RegisterRules(reg *acttypes.RulesRegistry) {
...
+		return k.actKeeper.GetRule(ctx, key.ApproveRuleId)
+	} else if space.ApproveSignRuleId > 0 {
+		return k.actKeeper.GetRule(ctx, space.ApproveSignRuleId)
+	} else {
+		return space.RuleApproveUpdateKey(), nil
+	}
+}
+
+func (k Keeper) getRejectUpdateKeyRule(ctx context.Context, space v1beta3.Space, key v1beta3.Key) (acttypes.Rule, error) {
+	if key.ApproveRuleId > 0 {
```
Comment (backsapc) [inline warden/x/warden/keeper/rules.go]: RejectRuleId

File: warden/x/warden/keeper/rules.go [hunk 13]
```diff
@@ -19,114 +19,244 @@ func (k Keeper) RegisterRules(reg *acttypes.RulesRegistry) {
...
-		return acttypes.Rule{}, err
+		return acttypes.Rule{}, acttypes.Rule{}, err
 	}
 
-	if space.SignRuleId > 0 {
-		return k.actKeeper.GetRule(ctx, space.SignRuleId)
-	} else {
-		return space.RuleNewKeyRequest(), nil
+	approveRule, err := k.getApproveNewKeyRequestRule(ctx, space)
+
```
Comment (backsapc) [inline warden/x/warden/keeper/rules.go]: Excess blank line

File: warden/x/warden/keeper/rules.go [hunk 14]
```diff
@@ -19,114 +19,244 @@ func (k Keeper) RegisterRules(reg *acttypes.RulesRegistry) {
...
+	if err != nil {
+		return acttypes.Rule{}, acttypes.Rule{}, err
+	}
+
+	rejectRule, err := k.getRejectNewKeyRequestRule(ctx, space)
+
```
Comment (backsapc) [inline warden/x/warden/keeper/rules.go]: Excess blank line

File: warden/x/warden/keeper/rules.go (hunk 6 above)
Comment (pitasi) [reply warden/x/warden/keeper/rules.go:125]: where is `getApproveKeyRule` defined?

Comment (mn13) [reply warden/x/warden/keeper/rules.go:125]: it was renamed to getApproveSignRule

**Review by backsapc — APPROVED** [review]

**Review by artur-abliazimov — APPROVED** [review]

**Pull request merged**